### PR TITLE
Pass the install action via settings to avoid issues with containers

### LINKED ISSUE #187: Installers blow up when used with Ninject
Version: Latest beta 6

### To repro

Use the `samples\nhibernate\custom-mappings\NHibernate_7` sample, install ninject, configure it to just ninject, start.

### Exception

```
Ninject.ActivationException was unhandled
  HResult=-2146233088
  Message=Error activating IntPtr
No matching bindings are available, and the type is not self-bindable.
Activation path:
  3) Injection of dependency IntPtr into parameter method of constructor of type Func{string, Task}
  2) Injection of dependency Func{string, Task} into parameter execute of constructor of type Installer+SchemaUpdater
  1) Request for Installer+SchemaUpdater

Suggestions:
  1) Ensure that you have defined a binding for IntPtr.
  2) If the binding was defined in a module, ensure that the module has been loaded into the kernel.
  3) Ensure you have not accidentally created more than one kernel.
  4) If you are using constructor arguments, ensure that the parameter name matches the constructors parameter name.
  5) If you are using automatic module loading, ensure the search path and filters are correct.

  Source=Ninject
  StackTrace:
       at Ninject.KernelBase.Resolve(IRequest request, Boolean handleMissingBindings)
       at Ninject.KernelBase.Resolve(IRequest request)
       at Ninject.Planning.Targets.Target`1.GetValue(Type service, IContext parent)
       at Ninject.Planning.Targets.Target`1.ResolveWithin(IContext parent)
       at Ninject.Activation.Providers.StandardProvider.GetValue(IContext context, ITarget target)
       at Ninject.Activation.Providers.StandardProvider.<>c__DisplayClass5.<Create>b__2(ITarget target)
       at System.Linq.Enumerable.WhereSelectArrayIterator`2.MoveNext()
       at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
       at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
       at Ninject.Activation.Providers.StandardProvider.Create(IContext context)
       at Ninject.Activation.Context.ResolveInternal(Object scope)
       at Ninject.Activation.Context.Resolve()
       at Ninject.KernelBase.Resolve(IRequest request, Boolean handleMissingBindings)
       at Ninject.KernelBase.Resolve(IRequest request, Boolean handleMissingBindings)
       at Ninject.KernelBase.Resolve(IRequest request)
       at Ninject.Planning.Targets.Target`1.GetValue(Type service, IContext parent)
       at Ninject.Planning.Targets.Target`1.ResolveWithin(IContext parent)
       at Ninject.Activation.Providers.StandardProvider.GetValue(IContext context, ITarget target)
       at Ninject.Activation.Providers.StandardProvider.<>c__DisplayClass5.<Create>b__2(ITarget target)
       at System.Linq.Enumerable.WhereSelectArrayIterator`2.MoveNext()
       at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
       at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
       at Ninject.Activation.Providers.StandardProvider.Create(IContext context)
       at Ninject.Activation.Context.ResolveInternal(Object scope)
       at Ninject.Activation.Context.Resolve()
       at Ninject.KernelBase.<>c__DisplayClass13.<Resolve>b__d(IBinding binding)
       at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
       at System.Linq.Enumerable.<CastIterator>d__94`1.MoveNext()
       at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source)
       at NServiceBus.Deduplication.NHibernate.Installer.Installer..ctor(IBuilder builder) in C:\BuildAgent\work\5135de308b2f3016\src\NServiceBus.NHibernate\Deduplication\Installer\Installer.cs:line 15
       at DynamicInjectorc882e4f4fdfa42dfab8af95361ecdb67(Object[] )
       at Ninject.Activation.Providers.StandardProvider.Create(IContext context)
       at Ninject.Activation.Context.ResolveInternal(Object scope)
       at Ninject.Activation.Context.Resolve()
       at Ninject.KernelBase.<>c__DisplayClass13.<Resolve>b__d(IBinding binding)
       at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
       at System.Linq.Enumerable.<CastIterator>d__94`1.MoveNext()
       at NServiceBus.StartableEndpoint.<RunInstallers>d__5.MoveNext() in C:\Build\src\NServiceBus.Core\StartableEndpoint.cs:line 93
    --- End of stack trace from previous location where exception was thrown ---
       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
       at NServiceBus.StartableEndpoint.<Start>d__1.MoveNext() in C:\Build\src\NServiceBus.Core\StartableEndpoint.cs:line 43
    --- End of stack trace from previous location where exception was thrown ---
       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
       at NServiceBus.Endpoint.<Start>d__1.MoveNext() in C:\Build\src\NServiceBus.Core\Endpoint.cs:line 28
    --- End of stack trace from previous location where exception was thrown ---
       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
       at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
       at Program.<AsyncMain>d__1.MoveNext() in C:\dev\docs.particular.net\samples\nhibernate\custom-mappings\NHibernate_7\Sample.XmlMapping\Program.cs:line 39
    --- End of stack trace from previous location where exception was thrown ---
       at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
       at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
       at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
       at Program.Main() in C:\dev\docs.particular.net\samples\nhibernate\custom-mappings\NHibernate_7\Sample.XmlMapping\Program.cs:line 14
  InnerException: 

```

### PR DESCRIPTION
**Pull request by andreasohlund**
Fixes #187 

Tested with the nhibernate sample

### CODE REVIEW

**Comment by andreasohlund** [comment]
@Particular/nservicebus-maintainers please review

Once in I propose we release since there is at least one beta user blocked by this one

File: src/NServiceBus.NHibernate/Deduplication/NHibernateGatewayDeduplication.cs [hunk 1]
```diff
@@ -19,6 +18,10 @@ public class NHibernateGatewayDeduplication : Feature
         public NHibernateGatewayDeduplication()
         {
             DependsOn("Gateway");
+
+            // since the installers are registered even if the feature isn't enabled we need to make 
+            // this a no-op of there is no "schema updater" available
+            Defaults(c=>c.Set<Installer.SchemaUpdater>(new Installer.SchemaUpdater()));
```
Comment (danielmarbach) [inline src/NServiceBus.NHibernate/Deduplication/NHibernateGatewayDeduplication.cs]: Shouldn't there be a spacing between the lambda?

Comment (danielmarbach) [inline src/NServiceBus.NHibernate/Deduplication/NHibernateGatewayDeduplication.cs]: I think the generics are redundant, run a cleanup?

File: src/NServiceBus.NHibernate/SynchronizedStorage/Installer/Installer.cs [hunk 2]
```diff
@@ -1,40 +1,32 @@
...
 
     class Installer : INeedToInstallSomething
     {
-        public Installer(IBuilder builder)
+        public Installer(ReadOnlySettings settings)
         {
-            // since the installers are registered even if the feature isn't enabled we need to make 
-            // this a no-op of there is no "schema updater" available 
-            schemaUpdater = builder.BuildAll<SchemaUpdater>().FirstOrDefault();
+            schemaUpdater = settings.Get<Unicast.Subscriptions.NHibernate.Installer.Installer.SchemaUpdater>();
```
Comment (danielmarbach) [inline src/NServiceBus.NHibernate/SynchronizedStorage/Installer/Installer.cs]: Does this need to be the fully qualified namespace?

File: src/NServiceBus.NHibernate/SynchronizedStorage/Installer/Installer.cs [hunk 3]
```diff
@@ -1,40 +1,32 @@
...
-            if (schemaUpdater == null)
-            {
-                return Task.FromResult(0);
-            }
-
             return schemaUpdater.Execute(identity);
         }
 
-        SchemaUpdater schemaUpdater;
+        Unicast.Subscriptions.NHibernate.Installer.Installer.SchemaUpdater schemaUpdater;
```
Comment (danielmarbach) [inline src/NServiceBus.NHibernate/SynchronizedStorage/Installer/Installer.cs]: Hu?

File: src/NServiceBus.NHibernate/Subscriptions/NHibernateSubscriptionStorage.cs [hunk 4]
```diff
@@ -34,20 +38,16 @@ protected override void Setup(FeatureConfigurationContext context)
             builder.AddMappings<SubscriptionMap>();
             var config = builder.Build();
 
-            Func<string, Task> installAction = _ => Task.FromResult(0);
-
             if (RunInstaller(context))
             {
-                installAction = identity =>
+                context.Settings.Get<Installer.SchemaUpdater>().Execute = identity =>
```
Comment (danielmarbach) [inline src/NServiceBus.NHibernate/Subscriptions/NHibernateSubscriptionStorage.cs]: Not a big fan of the Execute setter and then overriding it. But I see it is a necessary evil. The other solution would be to create a default empty one in the settings and overwrite it with a new instance of an Updater when necessary. I would personally prefer the later approach

File: src/NServiceBus.NHibernate/SynchronizedStorage/Installer/Installer.cs (hunk 3 above)
Comment (andreasohlund) [inline src/NServiceBus.NHibernate/SynchronizedStorage/Installer/Installer.cs]: Misstake, will fix

File: src/NServiceBus.NHibernate/Subscriptions/NHibernateSubscriptionStorage.cs (hunk 4 above)
Comment (andreasohlund) [inline src/NServiceBus.NHibernate/Subscriptions/NHibernateSubscriptionStorage.cs]: Setting are readonly here so I have to update an existing item

File: src/NServiceBus.NHibernate/Deduplication/NHibernateGatewayDeduplication.cs (hunk 1 above)
Comment (andreasohlund) [inline src/NServiceBus.NHibernate/Deduplication/NHibernateGatewayDeduplication.cs]: Set requires a type (not sure why but it does)

File: src/NServiceBus.NHibernate/SynchronizedStorage/Installer/Installer.cs (hunk 2 above)
Comment (andreasohlund) [inline src/NServiceBus.NHibernate/SynchronizedStorage/Installer/Installer.cs]: Misstake (good catch), will fix

**Comment by timbussmann** [comment]
> @Particular/nservicebus-maintainers please review

didn't you mean @Particular/nhibernate-persistence-maintainers ?

**Comment by andreasohlund** [comment]
>didn't you mean @Particular/nhibernate-persistence-maintainers ?

Yes I did :)

**Comment by andreasohlund** [comment]
@danielmarbach the actionable comments have been adressed.

@Particular/nhibernate-persistence-maintainers please review

**Comment by andreasohlund** [comment]
@Particular/nhibernate-persistence-maintainers please review and merge, we have a beta user blocked by this so I want to release a new beta

**Comment by HEskandari** [comment]
@andreasohlund I have...did we decide on Execute setter? could the setter be made internal at least?

**Comment by HEskandari** [comment]
Scratch that...looks to be an inner class of an internal, so it's okay. I'm happy with the change, anyone else needs to review? @Particular/nhibernate-persistence-maintainers

**Comment by DavidBoike** [comment]
:+1: from me. I used to be a persistence maintainer once. ;-)

Looks like Raven isn't the only maintainer group that needs some more members.

**Comment by HEskandari** [comment]
Merged. I'll release today.

**Comment by HEskandari** [comment]
@andreasohlund oh snap...this broke bunch of acceptance tests!

**Comment by HEskandari** [comment]
Just to be clear, the tests (NServiceBus.NHibernate.AcceptanceTests) are failing when you run them locally. Here's the exception:

    System.AggregateException : Test run failed due to one or more exceptions
      ----> NServiceBus.AcceptanceTesting.Support.ScenarioException : Endpoint failed to start
      ----> System.Exception : The given key (NServiceBus.TimeoutPersisters.NHibernate.Installer.Installer+SchemaUpdater) was not present in the dictionary.
      at NServiceBus.AcceptanceTesting.Support.ScenarioRunner.<Run>d__0.MoveNext() in C:\Build\src\NServiceBus.AcceptanceTesting\Support\ScenarioRunner.cs:line 88

**Comment by andreasohlund** [comment]
I'll look into it

**Comment by andreasohlund** [comment]
Found the issue, new PR here: https://github.com/Particular/NServiceBus.NHibernate/pull/192

**Pull request merged**